man/queue/sp_queue_print.3
.\"  Staple - A general-purpose data structure library in pure C89.
.\"  Copyright (C) 2021  Randoragon
.\"
.\"  This library is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU Lesser General Public
.\"  License as published by the Free Software Foundation;
.\"  version 2.1 of the License.
.\"
.\"  This library is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"  Lesser General Public License for more details.
.\"
.\"  You should have received a copy of the GNU Lesser General Public
.\"  License along with this library; if not, write to the Free Software
.\"  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
.\"--------------------------------------------------------------------------------
.TH SP_QUEUE_PRINT 3 DATE "libstaple-VERSION"
.SH NAME
sp_queue_print,
sp_queue_printc,
sp_queue_prints,
sp_queue_printi,
sp_queue_printl,
sp_queue_printsc,
sp_queue_printuc,
sp_queue_printus,
sp_queue_printui,
sp_queue_printul,
sp_queue_printf,
sp_queue_printd,
sp_queue_printld,
sp_queue_printll,
sp_queue_printull,
sp_queue_printu8,
sp_queue_printu16,
sp_queue_printu32,
sp_queue_printu64,
sp_queue_printi8,
sp_queue_printi16,
sp_queue_printi32,
sp_queue_printi64,
sp_queue_printstr
\- print contents of a queue
.SH SYNOPSIS
.ad l
#include <staple.h>
.sp
int
.BR sp_queue_print "(const struct sp_queue"
.RI * queue ,
int
.RI (* func ")(const void*))"
.br
int
.BR sp_queue_printc "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_prints "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printi "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printl "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printsc "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printuc "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printus "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printui "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printul "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printf "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printd "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printld "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printll "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printull "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printu8 "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printu16 "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printu32 "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printu64 "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printi8 "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printi16 "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printi32 "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printi64 "(const struct sp_queue"
.RI * queue )
.br
int
.BR sp_queue_printstr "(const struct sp_queue"
.RI * queue )
.sp
Link with \fI-lstaple\fP.
.ad
.SH DESCRIPTION
.P
The functions in the
.BR sp_queue_print ()
family are used to print useful information about a queue, and are
provided for debugging purposes.
.P
.I queue
is the queue that shall be printed.
.SS Generic Form
.I func
is an optional pointer to a custom print function. The function must be
implemented in such a way that it accepts the address of
an element as its argument.
.P
The function pointed to by
.I func
must return 0 for success, and anything else for failure. In debug mode, the
caught error code (if any) is printed on stderr.
.P
.B NULL
can be passed as
.IR func ,
in which case
.BR sp_queue_print ()
will default to printing the address of each element.
.SH RETURN VALUE
If successful, the functions in the
.BR sp_queue_print ()
family shall return 0. Otherwise, they shall return a code indicating the
error.
.SH ERRORS
The functions in the
.BR sp_queue_print ()
family shall fail if:
.IP \fBSP_EINVAL\fP 1.5i
.I queue
is a null pointer (debug mode only).
.SS Generic Form
The
.BR sp_queue_print ()
function shall fail if:
.IP \fBSP_ECALLBK\fP 1.5i
.I func
returned non-0 for some queue element.
.SS Suffixed Form
The suffixed
.BR sp_queue_print ()
functions shall fail if:
.IP \fBSP_EILLEGAL\fP 1.5i
.IR queue->elem_size
does not match the size of the function's return type (debug mode only).
.SH CONFORMING TO
See
.BR libstaple (7).
.SH SEE ALSO
.ad l
.BR libstaple (7),
.BR sp_queue (7),
.BR sp_queue_create (3),
.BR sp_queue_destroy (3),
.BR sp_queue_clear (3),
.BR sp_queue_push (3),
.BR sp_queue_peek (3),
.BR sp_queue_pop (3),
.BR sp_queue_insert (3),
.BR sp_queue_remove (3),
.BR sp_queue_qinsert (3),
.BR sp_queue_qremove (3),
.BR sp_queue_get (3),
.BR sp_queue_set (3),
.BR sp_queue_eq (3),
.BR sp_queue_copy (3),
.BR sp_queue_map (3)


src/queue/sp_queue_print.c
/*  Staple - A general-purpose data structure library in pure C89.
 *  Copyright (C) 2021  Randoragon
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation;
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "../sp_queue.h"
#include "../internal.h"
#include "../sp_errcodes.h"

int sp_queue_print(const struct sp_queue *queue, int (*func)(const void*))
{
	size_t i;
#ifdef STAPLE_DEBUG
	if (queue == NULL) {
		error(("queue is NULL"));
		return SP_EINVAL;
	}
#endif
	printf("sp_queue_print()\nsize/capacity: "SP_SIZE_FMT"/"SP_SIZE_FMT", elem_size: "SP_SIZE_FMT"\n",
		(SP_SIZE_T)queue->size, (SP_SIZE_T)queue->capacity, (SP_SIZE_T)queue->elem_size);
	if (func == NULL)
		for (i = 0; i < queue->size; i++) {
			const void *const elem = sp_ringbuf_get(i, queue->data, queue->capacity, queue->elem_size, queue->head);
			printf("["SP_SIZE_FMT"]\t%p\n", (SP_SIZE_T)i, elem);
		}
	else
		for (i = 0; i < queue->size; i++) {
			const void *const elem = sp_ringbuf_get(i, queue->data, queue->capacity, queue->elem_size, queue->head);
			int err;
			printf("["SP_SIZE_FMT"]\t", (SP_SIZE_T)i);
			if ((err = func(elem)) != 0) {
				error(("callback function func returned %d (non-0)", err));
				return SP_ECALLBK;
			}
		}
	return 0;
}
